include/bout/region.hxx

```diff
@@ -0,0 +1,236 @@
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * BOUT++ is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with BOUT++.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ **************************************************************************/
+
+#ifndef __REGION_H__
+#define __REGION_H__
+
+#include <type_traits>
+#include <vector>
+#include <utility>
+
+#include "bout/openmpwrap.hxx"
+
+/*
+ * The MAXREGIONBLOCKSIZE value can be tuned to try to optimise
+ * performance on specific hardware. It determines what the largest
+ * contiguous block size can be. As we hope the compiler will vectorise
+ * the access to these contiguous blocks, the optimal MAXREGIONBLOCKSIZE 
+ * is likely related to the vector size etc.
+ */
+#ifndef MAXREGIONBLOCKSIZE
+#define MAXREGIONBLOCKSIZE 64
+#endif
+#if MAXREGIONBLOCKSIZE <= 0
+#error "In region.hxx must set MAXREGIONBLOCKSIZE > 0"
+#endif
+
+//The following loops vectorise well well index is type int, needs testing
+//with current approach where index is ind2D or ind3D
+#define BLOCK_REGION_LOOP_SERIAL(region, index, ...)			\
+  {const auto blocks = region.getBlocks();				\
+  for(auto block = blocks.begin(); block < blocks.end() ; ++block){ \
+  for(auto index = (*block).first ; index <= (*block).second; ++index){ \
+  __VA_ARGS__ \
+    }}}
+
+#define BLOCK_REGION_LOOP(region, index, ...)			\
+  {const auto blocks = region.getBlocks();			\
+  BOUT_OMP(parallel for)				    \
+  for(auto block = blocks.begin(); block < blocks.end() ; ++block){ \
+  for(auto index = (*block).first ; index <= (*block).second; ++index){ \
+  __VA_ARGS__ \
+    }}}
+
+class specificInd{
+public:
+  int ind;
+  specificInd(){};
+  specificInd(int i): ind(i){};
+  specificInd& operator++(){++ind;return *this;}
+  specificInd& operator++(int){++ind;return *this;}
+  bool operator==(const specificInd &x) const {
+    return ind == x.ind;
+  }
+  bool operator!=(const specificInd &x) const {
+    return ind != x.ind;
+  }
+  bool operator>(const specificInd &x) const {
+    return ind > x.ind;
+  }
+  bool operator>=(const specificInd &x) const {
+    return ind >= x.ind;
+  }
+  bool operator<(const specificInd &x) const {
+    return ind < x.ind;
+  }
+  bool operator<=(const specificInd &x) const {
+    return ind <= x.ind;
+  }
+};
+//Make identical subclasses with different names
+class ind3D: public specificInd{
+public:
+  ind3D(): specificInd(){};
+  ind3D(int i): specificInd(i){};
+  //Note operator= from base class is always hidden
+  //by implicit method so have to be explicit
+  ind3D& operator=(const int& i){ind=i;return *this;}
```

Comment:
why take an int as const ref?
Taking by value should be faster.

Reply:
No good reason.